api/provider/aws/job_runner.go

```diff
@@ -1,25 +1,256 @@
 package aws
 
 import (
+	"encoding/json"
 	"fmt"
+	"time"
 
+	"github.com/aws/aws-sdk-go/aws/awserr"
 	"github.com/quintilesims/layer0/api/job"
+	"github.com/quintilesims/layer0/api/provider"
+	"github.com/quintilesims/layer0/common/errors"
 	"github.com/quintilesims/layer0/common/models"
 )
 
-func NewJobRunner(jobStore job.Store) job.RunnerFunc {
-	return job.RunnerFunc(func(j models.Job) error {
-		switch job.JobType(j.Type) {
-		case job.DeleteEnvironmentJob:
-			return deleteEnvironmentRunner(jobStore, j)
-		default:
-			return fmt.Errorf("Unrecognized JobType '%s'", j.Type)
+type JobRunner struct {
+	deployProvider       provider.DeployProvider
+	environmentProvider  provider.EnvironmentProvider
+	loadBalancerProvider provider.LoadBalancerProvider
+	serviceProvider      provider.ServiceProvider
+	taskProvider         provider.TaskProvider
+	jobStore             job.Store
+}
+
+func NewJobRunner(
+	d provider.DeployProvider,
+	e provider.EnvironmentProvider,
+	l provider.LoadBalancerProvider,
+	s provider.ServiceProvider,
+	t provider.TaskProvider,
+	store job.Store,
+) *JobRunner {
+	return &JobRunner{
+		deployProvider:       d,
+		environmentProvider:  e,
+		loadBalancerProvider: l,
+		serviceProvider:      s,
+		taskProvider:         t,
+		jobStore:             store,
+	}
+}
+
+func (r *JobRunner) Run(j models.Job) error {
+	switch job.JobType(j.Type) {
+	case job.CreateDeployJob:
+		return r.createDeploy(j.JobID, j.Request)
+	case job.CreateEnvironmentJob:
+		return r.createEnvironment(j.JobID, j.Request)
+	case job.CreateLoadBalancerJob:
+		return r.createLoadBalancer(j.JobID, j.Request)
+	case job.CreateServiceJob:
+		return r.createService(j.JobID, j.Request)
+	case job.CreateTaskJob:
+		return r.createTask(j.JobID, j.Request)
+	case job.DeleteDeployJob:
+		return r.deleteDeploy(j.JobID, j.Request)
+	case job.DeleteEnvironmentJob:
+		return r.deleteEnvironment(j.JobID, j.Request)
+	case job.DeleteLoadBalancerJob:
+		return r.deleteLoadBalancer(j.JobID, j.Request)
+	case job.DeleteServiceJob:
+		return r.deleteService(j.JobID, j.Request)
+	case job.DeleteTaskJob:
+		return r.deleteTask(j.JobID, j.Request)
+	case job.UpdateEnvironmentJob:
+		return r.updateEnvironment(j.JobID, j.Request)
+	case job.UpdateLoadBalancerJob:
+		return r.updateLoadBalancer(j.JobID, j.Request)
+	case job.UpdateServiceJob:
+		return r.updateService(j.JobID, j.Request)
+	default:
+		return fmt.Errorf("Unrecognized JobType '%s'", j.Type)
+	}
+}
+
+func (r *JobRunner) createDeploy(jobID, request string) error {
+	var req models.CreateDeployRequest
+	if err := json.Unmarshal([]byte(request), &req); err != nil {
+		return errors.New(errors.InvalidRequest, err)
+	}
+
+	deploy, err := r.deployProvider.Create(req)
+	if err != nil {
+		return err
```

Comment:
nope, job worker is responsible for that.

Reply:
we could change the job runner signature to return `(string, error)`  instead of just `error`. The worker would know to set the `JobStatus` to whatever string was returned. That could help make it clearer which part is responsible for updating the job store.